src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -0,0 +1,187 @@
+﻿using System;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	internal sealed class ImmutableDefinitionChecker {
+		private readonly SemanticModel m_model;
+		private readonly Action<Diagnostic> m_diagnosticSink;
+		private readonly ImmutabilityContext m_context;
+
+		public ImmutableDefinitionChecker(
+			SemanticModel model,
+			Action<Diagnostic> diagnosticSink,

```

Comment:
Why this diagnosticSink stuff?

Having nested generators made the stuff uglier.

Callers would just pass `m_ctx.ReportDiagnostic` here.